### PR TITLE
GH-98522: Add version number to code objects, to provide better version numbers for closures and comprehensions.

### PR DESCRIPTION
**Pull request by markshannon**
Measured performance impact [might be positive](https://gist.github.com/markshannon/181ac95ce5b543ba5cafd43677829266) but is too close to noise to be meaningful.

* Issue: gh-98522

### CODE REVIEW

File: Tools/build/deepfreeze.py [hunk 1]
```diff
@@ -270,6 +272,8 @@ def generate_code(self, name: str, code: types.CodeType) -> str:
             self.write(f".co_nplaincellvars = {nplaincellvars},")
             self.write(f".co_ncellvars = {ncellvars},")
             self.write(f".co_nfreevars = {nfreevars},")
+            self.write(f".co_version = {next_code_version},")
+            next_code_version += 1
```
Comment (iritkatriel) [inline Tools/build/deepfreeze.py:274]: assert that it's less than 10000, or is this crazy?

Comment (markshannon) [reply Tools/build/deepfreeze.py:274]: The 10000 is just a hack. I'll fix it properly.

File: Tools/build/deepfreeze.py [hunk 2]
```diff
@@ -461,6 +465,7 @@ def generate(args: list[str], output: TextIO) -> None:
             with printer.block(f"if ({p} < 0)"):
                 printer.write("return -1;")
         printer.write("return 0;")
+    printer.write(f"\nuint32_t _Py_next_func_version = {next_code_version};\n")
```
Comment (iritkatriel) [inline Tools/build/deepfreeze.py:468]: does it need to be incremented here?

File: Include/internal/pycore_code.h [hunk 3]
```diff
@@ -472,6 +472,8 @@ typedef struct _PyShimCodeDef {
 extern PyCodeObject *
 _Py_MakeShimCode(const _PyShimCodeDef *code);
 
+extern uint32_t _Py_next_func_version;
```
Comment (iritkatriel) [inline Include/internal/pycore_code.h:477]: Is there a reason why this is called func_version and not code_version? I find it a bit confusing (note that in deepfreeze.py we now have ``_Py_next_func_version = {next_code_version}``).

File: Tools/build/deepfreeze.py (hunk 2 above)
Comment (markshannon) [reply Tools/build/deepfreeze.py:468]: No, as it is the next version. It gets incremented when used.

File: Include/internal/pycore_code.h (hunk 3 above)
Comment (markshannon) [reply Include/internal/pycore_code.h:477]: It is primarily used as the version for functions.
https://github.com/python/cpython/blob/main/Python/bytecodes.c#L2976
Caching it in the code object is a further optimization.

**Review by iritkatriel — APPROVED** [review]

**Comment by markshannon** [comment]
Address sanitizer test failing itself because its too slow :raised_eyebrow:

**Pull request merged**